File: gosi-kafka-sdk-dotnet/Auth/MutualTlsAuthHandler.cs
namespace Gosi.Kafka.Sdk.Auth;

using System;
using Confluent.Kafka;

/// <summary>
/// Configures Mutual TLS (mTLS) authentication for .NET clients.
/// </summary>
public class MutualTlsAuthHandler : IAuthenticationHandler
{
    private readonly string _keystoreLocation;
    private readonly string _keystorePassword;
    private readonly string _truststoreLocation;

    public MutualTlsAuthHandler(string keystoreLocation, string keystorePassword, string truststoreLocation)
    {
        if (string.IsNullOrWhiteSpace(keystoreLocation)) throw new ArgumentException("Keystore location cannot be empty");

        _keystoreLocation = keystoreLocation;
        _keystorePassword = keystorePassword;
        _truststoreLocation = truststoreLocation;
    }

    public void Configure(ClientConfig config)
    {
        config.SecurityProtocol = SecurityProtocol.Ssl;
        config.SslKeystoreLocation = _keystoreLocation;
        config.SslKeystorePassword = _keystorePassword;
        
        if (!string.IsNullOrWhiteSpace(_truststoreLocation))
        {
            config.SslCaLocation = _truststoreLocation;
        }
    }
}

File: gosi-kafka-sdk-java/gosi-kafka-sdk-core/src/main/java/com/gosi/kafka/sdk/auth/MutualTlsAuthHandler.java
package com.gosi.kafka.sdk.auth;

import org.apache.kafka.clients.CommonClientConfigs;
import org.apache.kafka.common.config.SslConfigs;

import java.util.Map;

/**
 * Configures Mutual TLS (mTLS) authentication.
 */
public class MutualTlsAuthHandler implements AuthenticationHandler {

    private final String keystoreLocation;
    private final String keystorePassword;
    private final String keyPassword;
    private final String truststoreLocation;
    private final String truststorePassword;

    public MutualTlsAuthHandler(String keystoreLocation, String keystorePassword, String keyPassword,
                                String truststoreLocation, String truststorePassword) {
        if (keystoreLocation == null || keystoreLocation.trim().isEmpty()) {
            throw new IllegalArgumentException("Keystore location cannot be empty");
        }
        this.keystoreLocation = keystoreLocation;
        this.keystorePassword = keystorePassword;
        this.keyPassword = keyPassword != null ? keyPassword : keystorePassword;
        this.truststoreLocation = truststoreLocation;
        this.truststorePassword = truststorePassword;
    }

    @Override
    public void configure(Map<String, Object> kafkaProperties) {
        kafkaProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SSL");
        
        kafkaProperties.put(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, keystoreLocation);
        if (keystorePassword != null) {
            kafkaProperties.put(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, keystorePassword);
        }
        if (keyPassword != null) {
            kafkaProperties.put(SslConfigs.SSL_KEY_PASSWORD_CONFIG, keyPassword);
        }
        
        if (truststoreLocation != null && !truststoreLocation.trim().isEmpty()) {
            kafkaProperties.put(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, truststoreLocation);
            if (truststorePassword != null) {
                kafkaProperties.put(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, truststorePassword);
            }
        }
    }
}
